pygls_testsuite/conftest.py

```diff
@@ -43,79 +48,140 @@
 DOC_URI = uris.from_fs_path(__file__)
 
 
-def pytest_sessionfinish(session, exitstatus):
-    """ whole test run finishes. """
-    sleep(CALL_TIMEOUT)
-    for cs in ClientServer._to_stop:
-       cs._stop()
-
-
-class ClientServer:
-    _to_stop: List['ClientServer'] = []
-
-    def __init__(self):
-        # Client to Server pipe
-        csr, csw = os.pipe()
-        # Server to client pipe
-        scr, scw = os.pipe()
-
-        # Setup Server
-        self.server = LanguageServer()
-        self.server_thread = Thread(target=self.server.start_io, args=(
-            os.fdopen(csr, 'rb'), os.fdopen(scw, 'wb')
-        ))
-        self.server_thread.daemon = True
-
-        # Setup client
-        self.client = LanguageServer(asyncio.new_event_loop())
-        self.client_thread = Thread(target=self.client.start_io, args=(
-            os.fdopen(scr, 'rb'), os.fdopen(csw, 'wb')))
-        self.client_thread.daemon = True
-
-    def start(self):
-        self.server_thread.start()
-        self.server.thread_id = self.server_thread.ident
-
-        self.client_thread.start()
-
-        self.initialize()
-
-    def stop(self):
-        ClientServer._to_stop.append(self)
-
-    def _stop(self):
-        shutdown_response = (
-            self.client
-            .lsp.send_request(pygls.lsp.methods.SHUTDOWN)
-            .result(timeout=CALL_TIMEOUT)
-        )
-        assert shutdown_response is None
-        # Exit the server
-        self.client.lsp.notify(pygls.lsp.methods.EXIT)
-        self.server_thread.join()
-        # Exit the client
-        self.client._stop_event.set()
-        try:
-            self.client.loop._signal_handlers.clear() # HACK ?
-        except AttributeError:
-            pass
-        self.client_thread.join()
-
-    def initialize(self):
-        response = self.client.lsp.send_request(
-            INITIALIZE,
-            InitializeParams(
-                process_id=12345,
-                root_uri="file://",
-                capabilities=ClientCapabilities()
+if IS_PYODIDE:
+    class ClientServer:
+
+        def __init__(self):
+
+            self.server = LanguageServer()
+            self.client = Mock()
+
+            # Pyodide specific implementations/patches.
+            self.server.lsp._send_data = self._send_data
+            self.client.lsp.send_request = self._send_request
```

Comment:
This overrides the [usual `send_request` implementation](https://github.com/openlawlibrary/pygls/blob/master/pygls/protocol.py#L477) with [this one](https://github.com/openlawlibrary/pygls/pull/239/commits/1de4efd2139b5663a22ba1a12a52c10abd2aa0bb#diff-aee6bad608e2e8758dbcab42e829eba66f0f394079ef176817844829c7617400R85) that passes the message direct to the server's `_procedure_handler`